### PR TITLE
feat: Add flush

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Add blocking flush to send out all envelope items.

This PR is based on https://github.com/getsentry/sentry-cocoa/pull/2169.

## :bulb: Motivation and Context

Fixes GH-1013

## :green_heart: How did you test it?

Unit tests and real devices.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

### CODE REVIEW

File: Tests/SentryTests/Networking/SentryHttpTransportTests.swift [hunk 1]
```diff
@@ -614,6 +619,72 @@ class SentryHttpTransportTests: XCTestCase {
...
+    func testFlush_BlocksCallingThread() {
+        CurrentDate.setCurrentDateProvider(DefaultCurrentDateProvider.sharedInstance())
+        
+        givenCachedEvents()
+        
+        fixture.requestManager.responseDelay = fixture.flushTimeout + 0.2
+        
+        let beforeFlush = Date()
+        let success = sut.flush(timeout: fixture.flushTimeout)
+        let blockingDuration = Date().timeIntervalSince(beforeFlush)
```
Comment (philipphofmann) [inline Tests/SentryTests/Networking/SentryHttpTransportTests.swift:632]: @bruno-garcia or @mattjohnsonpint, do you have any suggestions on how to properly test if a call to a function is blocking?

Comment (mattjohnsonpint) [reply Tests/SentryTests/Networking/SentryHttpTransportTests.swift]: We have a [similar test in .NET](https://github.com/getsentry/sentry-dotnet/blob/cb735d1f62e062b6cf1ef3e75f22e914f0c98668/test/Sentry.Tests/Internals/BackgroundWorkerTests.cs#L361-L384), where we also just test that the flush timeout period has elapsed.

We do the blocking slightly differently, by using an internal `process: false` flag to the background worker, which just tells it not to process items after enqueueing them.  I believe a previous version of the test took an approach similar to yours where we just added an extra long delay to the transport.  I can't recall why we switched approaches, but I think for this particular case either would work.

Comment (mattjohnsonpint) [reply Tests/SentryTests/Networking/SentryHttpTransportTests.swift]: My only suggestion would be to use a monotonic timing source, but I think you already made that change.

Comment (philipphofmann) [reply Tests/SentryTests/Networking/SentryHttpTransportTests.swift]: Thanks, @mattjohnsonpint.

**Review by brustolin — APPROVED** [review]
LGTM. 

I have only one observation regarding comments.

File: Sources/Sentry/Public/SentryClient.h [hunk 2]
```diff
@@ -113,6 +113,15 @@ SENTRY_NO_INIT
 
 - (void)captureEnvelope:(SentryEnvelope *)envelope NS_SWIFT_NAME(capture(envelope:));
 
+/**
+ * Waits synchronously for the SDK to flush out all queued and cached items. Returns if the SDK
+ * doesn't complete the flush before the specified timeout period has elapsed or if there is no
+ * internet connection. The SDK doesn't dispose the client or the hub.
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/Public/SentryClient.h]: Thanks for the feedback. Is it better now?

**Pull request merged**